source/vstd/std_specs/hash.rs

```diff
@@ -0,0 +1,383 @@
+// where you'd expect to find `key: &Key`. This allows for the case
+// that `Key` can be borrowed as something other than `&Key`. For
+// instance, `Box<u32>` can be borrowed as `&u32` and `String` can be
+// borrowed as `&str`, so in those cases `Q` would be `u32` and `str`
+// respectively.
+// To deal with this, we have a specification function that opaquely
+// specifies what it means for a map to contain a borrowed key of type
+// `&Q`. And the postcondition of `contains_key` just says that its
+// result matches the output of that specification function. But this
+// isn't very helpful by itself, since there's no body to that
+// specification function. So we have special-case axioms that say
+// what this means in two important circumstances: (1) `Key = Q` and
+// (2) `Key = Box<Q>`. (TODO: Include another axiom for the case `Key
+// = String, Q = str`.)
+pub spec fn map_contains_borrowed_key<Key, Value, Q: ?Sized>(m: Map<Key, Value>, k: &Q) -> bool;
+
+pub broadcast proof fn axiom_hash_map_contains_deref_key<Q, Value>(m: Map<Q, Value>, k: &Q)
+    ensures
+        #[trigger] map_contains_borrowed_key::<Q, Value, Q>(m, k) <==> m.contains_key(*k),
+{
+    admit();
+}
+
+pub broadcast proof fn axiom_hash_map_contains_box<Q, Value>(m: Map<Box<Q>, Value>, k: &Q)
+    ensures
+        #[trigger] map_contains_borrowed_key::<Box<Q>, Value, Q>(m, k) <==> m.contains_key(
+            Box::new(*k),
+        ),
+{
+    admit();
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_map_contains_key<Key, Value, S, Q>(m: &HashMap<Key, Value, S>, k: &Q) -> (result:
+    bool) where Key: Borrow<Q> + Hash + Eq, Q: Hash + Eq + ?Sized, S: BuildHasher
+    ensures
+        does_type_have_deterministic_hash::<Key>() && does_type_conform_to_build_hasher_model::<S>()
+            ==> result == map_contains_borrowed_key(m@, k),
+{
+    m.contains_key(k)
+}
+
+// The specification for `get` has a parameter `key: &Q` where you'd
+// expect to find `key: &Key`. This allows for the case that `Key` can
+// be borrowed as something other than `&Key`. For instance,
+// `Box<u32>` can be borrowed as `&u32` and `String` can be borrowed
+// as `&str`, so in those cases `Q` would be `u32` and `str`
+// respectively.
+// To deal with this, we have a specification function that opaquely
+// specifies what it means for a map to map a borrowed key of type
+// `&Q` to a certain value. And the postcondition of `get` says that
+// its result matches the output of that specification function. (It
+// also says that its result corresponds to the output of
+// `map_contains_borrowed_key`, discussed above.) But this isn't very
+// helpful by itself, since there's no body to that specification
+// function. So we have special-case axioms that say what this means
+// in two important circumstances: (1) `Key = Q` and (2) `Key =
+// Box<Q>`. (TODO: Include another axiom for the case `Key = String, Q
+// = str`.)
+pub spec fn map_maps_borrowed_key_to_value<Key, Value, Q: ?Sized>(
```

Comment:
Can `map_maps_borrowed_key_to_value` be shortened to `maps_borrowed_key_to_value`?

Reply:
Sure, I'll do that.